src/filters/capture.rs

```diff
@@ -14,71 +14,73 @@
  * limitations under the License.
  */
 
-mod capture;
+mod affix;
 mod config;
 mod metrics;
-mod proto;
+mod regex;
 
-use std::sync::Arc;
+crate::include_proto!("quilkin.extensions.filters.capture.v1alpha1");
 
-use tracing::warn;
+use std::sync::Arc;
 
 use crate::{filters::prelude::*, metadata::Value};
 
-use capture::Capture;
-use metrics::Metrics;
-use proto::quilkin::extensions::filters::capture_bytes::v1alpha1::CaptureBytes as ProtoConfig;
+use self::{
+    affix::{Prefix, Suffix},
+    metrics::Metrics,
+    regex::Regex,
+};
 
+use self::quilkin::extensions::filters::capture::v1alpha1 as proto;
 pub use config::{Config, Strategy};
 
-pub const NAME: &str = "quilkin.extensions.filters.capture_bytes.v1alpha1.CaptureBytes";
+pub const NAME: &str = "quilkin.extensions.filters.capture.v1alpha1.Capture";
 
 /// Creates a new factory for generating capture filters.
 pub fn factory() -> DynFilterFactory {
     Box::from(CaptureBytesFactory::new())
 }
 
-struct CaptureBytes {
-    capture: Box<dyn Capture + Sync + Send>,
+/// Trait to implement different strategies for capturing packet data.
+pub trait CaptureStrategy {
+    /// Capture packet data from the contents, and optionally returns a value if
+    /// anything was captured.
+    fn capture(&self, contents: &mut Vec<u8>, metrics: &Metrics) -> Option<Value>;
+}
+
+struct Capture {
+    capture: Box<dyn CaptureStrategy + Sync + Send>,
     /// metrics reporter for this filter.
     metrics: Metrics,
     metadata_key: Arc<String>,
-    size: usize,
-    remove: bool,
 }
 
-impl CaptureBytes {
+impl Capture {
     fn new(config: Config, metrics: Metrics) -> Self {
-        CaptureBytes {
-            capture: config.strategy.as_capture(),
+        Self {
+            capture: config.strategy.into_capture(),
             metrics,
             metadata_key: Arc::new(config.metadata_key),
-            size: config.size,
-            remove: config.remove,
         }
     }
 }
 
-impl Filter for CaptureBytes {
+impl Filter for Capture {
     #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, ctx)))]
     fn read(&self, mut ctx: ReadContext) -> Option<ReadResponse> {
-        // if the capture size is bigger than the packet size, then we drop the packet,
-        // and occasionally warn
-        if ctx.contents.len() < self.size {
-            if self.metrics.packets_dropped_total.get() % 1000 == 0 {
-                warn!(count = ?self.metrics.packets_dropped_total.get(), "Packets are being dropped due to their length being less than {} bytes", self.size);
-            }
-            self.metrics.packets_dropped_total.inc();
-            return None;
-        }
-        let token = self
-            .capture
-            .capture(&mut ctx.contents, self.size, self.remove);
+        let capture = self.capture.capture(&mut ctx.contents, &self.metrics);
+        let mut capture_key = (*self.metadata_key).clone();
+        capture_key += "/is_present";
```

Comment:
the clone+concat looks expensive, we should be able to create the string once on instantiation since its not changing?

Can you elaborate what `/is_present` is for I didn't understand from the PR description and it wasnt clear looking at the code?

Reply:
`/is_present` is more useful with the inclusion of the `Match` filter and other conditional flow filters. I wrote an [example](https://github.com/googleforgames/quilkin/discussions/459#discussion-3764877) that uses it, but in essence, it allows you to check for whether a capture succeeded without checking the specific contents.

This is more useful with the inclusion of regex because something like `\d{8}$` would capture the last eight digits, and if you wanted to check if that worked you'd need to match every possible combination of those digits, where as   with `/is_present` we can check it with a couple lines. E.g.

```yaml
  filters:
  - name: quilkin.extensions.filters.capture.v1alpha1.Capture
    config:
      strategy:
        kind: regex
        regex: \d{8}$
  - name: quilkin.extensions.filters.matches.v1alpha1.Matches
    config:
        metadataKey: quilkin.dev/capture_bytes/is_present
        fallthrough: pass
        branches:
        # If present, drop the packet.
        - value: true
          filter: quilkin.extensions.filters.drop.v1alpha1.Drop
```